Switch logging to the tracing library

mcelrath opened the pull request:
Switch to the `tracing` library instead of `log` for systematic logging:
1. libp2p already uses it
2. Logs crate/module emitting the message
3. Replace all `println!()` statements with equivalent `tracing::info/error/debug` statements
4. Remove `env_logger` dependency
5. Removed `log` dependency in favor of `tracing-log` to intercept calls to `log::debug!` (etc) and forward them to the tracing module, for dependencies that use `log`.
6. Systematically use structured logging instead of string interpolation
7. Add `strum` as a dependency for better logging of errors involving enums (for `BitcoinRequest`, mostly in `ipc/client.rs`).
8. Propagate `template_id` and `block_hash` through the stratum and ipc interfaces for logging
9. Systematically use `template_id` as an integer instead of mixed string/int usage
10. Remove flexbuffers dependency (we removed it long ago) Fixes #165 
11. More clearly log Braid::extend and log failure to extend

Here's a list of direct dependencies that use `log`:
1. `sqlx` - Database connectivity (via sqlx-core, sqlx-sqlite)
2. `bitcoincore-rpc` - Bitcoin Core RPC client
3. `libp2p` - Peer-to-peer networking (via many sub-crates)
4. `jsonrpsee` - JSON-RPC client/server (via jsonrpsee-http-client)
5. `bitcoincore-zmq` - Bitcoin Core ZMQ bindings (via async_zmq → mio)

Here's a list of indirect dependencies that use `log`:
  - `async-io` - Used by sqlx-core
  - `async-std` - Used by sqlx
  - `kv-log-macro` - Used by async-std
  - `hyper-rustls` - HTTPS support for jsonrpsee
  - `if-watch` - Network interface monitoring for libp2p
  - `mio` (v0.6) - Async I/O for bitcoincore-zmq
  - `multistream-select` - Protocol negotiation for libp2p
  - `netlink-proto` & `netlink-sys` - Linux netlink for if-watch
  - `polling` - Polling mechanisms for async-io
  - `rtnetlink` - Linux routing for if-watch
  - `rustls` - TLS implementation (multiple dependents)
  - `rustls-platform-verifier` - Platform certificate verification
  - `soketto` - WebSocket for jsonrpsee
  - `tokio` - Core async runtime
  - `tracing-log` - The bridge itself (used by tracing-subscriber)
  - `tracing-subscriber` - tracing infrastructure

This PR is targeting the `15-10-25-db-integration` (#265) to minimize the diff. I will merge #265 first and change the target of this PR after merging it.

zaidmstrr reviewed (commented):
Looks good, things like `[instrument]` are really beneficial for debugging.

Sansh2356 reviewed (dismissed):
LGTM @mcelrath .

mcelrath committed:
Switch from log to tracing module

mcelrath committed:
refactor: migrate logging to structured tracing fields

    - Refactor logging statements across to use structured fields
    - Replace string interpolation with field-based logging
    - Add consistent field naming (peer, template_id, hash, error, etc.)

mcelrath committed:
More structured logging

mcelrath committed:
merkel -> merkle

mcelrath committed:
Even more structured logging

mcelrath committed:
More structured logging, log actual listening IPs for cut and paste

mcelrath committed:
Log miner connection_id with messages specific to it

mcelrath committed:
More clearly log listening endpoints

mcelrath committed:
Propagate template_id and log block hash in IPC code

    - use hex::encode() instead of bespoke bytes_to_hex
    - use BlockHeader everywhere instead of individual fields (makes
      calculating block hash easier and is more concise)

mcelrath committed:
Convert template_id to an integer everywhere

mcelrath committed:
Misc message updates in main, remove flexbuffers

mcelrath committed:
More concise logging of db load

mcelrath committed:
Make rustfmt happy

mcelrath committed:
Misc message logging

mcelrath committed:
More clearly log braid extend and tips

mcelrath committed:
Make rustmt happy

Sansh2356 commented on node/src/main.rs:
```diff
@@ -65,22 +66,25 @@ async fn main() -> Result<(), Box<dyn Error>> {
     //Reconstructing local braid upon startup
     let db_connection_pool_ref = _db_handler.db_connection_pool.clone();
     let braid_ref = braid.clone();
+    // FIXME instead we should look 144 blocks back from the bitcoin tip (1 day) and load beads
```
Can you elaborate further on this @mcelrath , since `IBD` is als currently structured to deal with no genesis bead .

mcelrath replied:
This is about how many beads we load into memory. One day seems like a reasonable number to keep in memory.

Also this block needs to be moved down, below where we check/create the data directory, because that's where braidpool.db lives. Presumably this barfs if ~/.braidpool hasn't been created yet?

Sansh2356 commented on node/src/stratum.rs:
`DownstreamClient` must have a `new` implementation apart from initializing a new client interface from `default`, just noting for a future reference .

mcelrath replied:
You mean instead of using `DownstreamClient::default()` and then modifying it, you want to use a `::new()` method? That's a good idea.

Sansh2356 commented on node/build.rs:
```diff
@@ -21,7 +21,7 @@ fn main() {
                         let path = entry.path();
                         if path.extension().and_then(|ext| ext.to_str()) == Some("capnp") {
-                            println!("cargo:warning=Compiling schema file: {:?}", path);
+                            println!("cargo:info=Compiling schema file: {:?}", path);
```
These logs are visible in the default `OUT_DIR` `env` variable  we can keep it at `warn` level also .

mcelrath replied:
The warning was just annoying me, and it seemed it was recompiling with every run of cargo, violating the configuration we have "cargo:rerun-if-changed". It doesn't seem to be doing that anymore though, I'll put it back to warn.

Sansh2356 commented on node/Cargo.toml:
```diff
@@ -35,7 +32,9 @@ toml = { workspace = true }
 async-trait = { workspace = true }
 sqlx = { workspace = true}
+strum = { workspace = true }
 anyhow = "1.0.100"
```
This dependency can be removed .

mcelrath replied:
This is used in ipc.rs to turn the `BitcoinRequest` enum into a string for logging purposes.

Sansh2356 commented on node/Cargo.toml:
```diff
@@ -35,7 +32,9 @@ toml = { workspace = true }
 sqlx = { workspace = true}
+strum = { workspace = true }
 anyhow = "1.0.100"
+if-addrs = "0.13"
```
Can we add this by inheriting from  `workspace` instead ?

Sansh2356 reviewed (dismissed):
ACK @mcelrath  [`2ec04dd`](https://github.com/braidpool/braidpool/pull/291/commits/2ec04dd56f013259ba3b6bde84238f23b86d883a), just few comments perhaps for discussion purposes apart from that can you squash some of the commits into a parent commit ?

mcelrath committed:
Restore cargo warn for capnp build, move deps to workspace

mcelrath commented:
> ACK @mcelrath [`2ec04dd`](https://github.com/braidpool/braidpool/pull/291/commits/2ec04dd56f013259ba3b6bde84238f23b86d883a), just few comments perhaps for discussion purposes apart from that can you squash some of the commits into a parent commit ?

I have github configured to do a squash & merge automatically.
